sdks/java/harness/src/test/java/org/apache/beam/fn/harness/BeamFnDataReadRunnerTest.java

```diff
@@ -122,334 +129,579 @@
-    assertEquals(expected, responseBuilder.build());
-
-    // Ensure that we process the correct number of elements after splitting.
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("A"));
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("B"));
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("C"));
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("D"));
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("E"));
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("F"));
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("G"));
-    assertThat(
-        outputValues,
-        contains(
-            valueInGlobalWindow("A"),
-            valueInGlobalWindow("B"),
-            valueInGlobalWindow("C"),
-            valueInGlobalWindow("D"),
-            valueInGlobalWindow("E")));
+    @Parameterized.Parameter(0)
+    public ProcessBundleSplitResponse expectedResponse;
+
+    @Parameterized.Parameter(1)
+    public long inputElements;
+
+    @Parameterized.Parameter(2)
+    public List<String> processedElements;
+
+    @Parameterized.Parameter(3)
+    public double fractionOfRemainder;
+
+    @Parameterized.Parameter(4)
+    public List<Long> allowedSplitPoints;
+
+    @Test
+    public void testChannelSplittingWithAllowedSplitPoints() throws Exception {
+      List<WindowedValue<String>> outputValues = new ArrayList<>();
+      BeamFnDataClient mockBeamFnDataClient = mock(BeamFnDataClient.class);
+      BeamFnDataReadRunner<String> readRunner =
+          createReadRunner(outputValues::add, PTRANSFORM_ID, mockBeamFnDataClient);
+      assertEquals(
+          expectedResponse,
+          executeSplit(
+              readRunner,
+              PTRANSFORM_ID,
+              inputElements,
+              processedElements,
+              fractionOfRemainder,
+              allowedSplitPoints));
+    }
   }
 
-  @Test
-  public void testSplittingWhenSomeElementsProcessed() throws Exception {
-    List<WindowedValue<String>> outputValues = new ArrayList<>();
-    BeamFnDataReadRunner<String> readRunner = createReadRunner(outputValues::add);
-
-    ProcessBundleSplitRequest request =
-        ProcessBundleSplitRequest.newBuilder()
-            .putDesiredSplits(
-                "pTransformId",
-                DesiredSplit.newBuilder()
-                    .setEstimatedInputElements(10)
-                    .setFractionOfRemainder(0.5)
-                    .build())
-            .build();
-    ProcessBundleSplitResponse.Builder responseBuilder = ProcessBundleSplitResponse.newBuilder();
-
-    // Process 2 elements then split
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("A"));
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("B"));
-    readRunner.trySplit(request, responseBuilder);
+  // Test different cases of element split with empty allowed split points.
+  @RunWith(Parameterized.class)
+  public static class ElementSplitTest {
```

Comment:
The same to [test_sdf_split](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/worker/bundle_processor_test.py#L101)